code/modules/integrated_electronics/core/assemblies.dm

```diff
@@ -0,0 +1,372 @@
+
+	return TRUE
+
+/obj/item/device/electronic_assembly/afterattack(atom/target, mob/user, proximity)
+	if(proximity)
+		var/scanned = FALSE
+		for(var/obj/item/integrated_circuit/input/sensor/S in contents)
+			S.set_pin_data(IC_OUTPUT, 1, weakref(target))
+			S.check_then_do_work()
+			if(S.scan(target))
+				scanned = TRUE
+		if(scanned)
+			visible_message("<span class='notice'>\The [user] waves \the [src] around [target].</span>")
+
+/obj/item/device/electronic_assembly/attackby(var/obj/item/I, var/mob/user)
+	if(istype(I, /obj/item/integrated_circuit))
+		if(!user.canUnEquip(I))
+			return FALSE
+		if(add_circuit(I, user))
+			to_chat(user, "<span class='notice'>You slide \the [I] inside \the [src].</span>")
+			playsound(get_turf(src), 'sound/items/Deconstruct.ogg', 50, 1)
+			interact(user)
+			return TRUE
+	else if(istype(I, /obj/item/crowbar))
+		playsound(get_turf(src), 'sound/items/Crowbar.ogg', 50, 1)
+		opened = !opened
+		to_chat(user, "<span class='notice'>You [opened ? "opened" : "closed"] \the [src].</span>")
+		update_icon()
+		return TRUE
+	else if(istype(I, /obj/item/device/integrated_electronics/wirer) || istype(I, /obj/item/device/integrated_electronics/debugger) || istype(I, /obj/item/screwdriver))
+		if(opened)
+			interact(user)
+		else
+			to_chat(user, "<span class='warning'>\The [src] isn't opened, so you can't fiddle with the internal components.  \
+			Try using a crowbar.</span>")
+	else if(istype(I, /obj/item/stock_parts/cell))
+		if(!opened)
+			to_chat(user, "<span class='warning'>\The [src] isn't opened, so you can't put anything inside.  Try using a crowbar.</span>")
+			return FALSE
+		if(battery)
+			to_chat(user, "<span class='warning'>\The [src] already has \a [battery] inside.  Remove it first if you want to replace it.</span>")
+			return FALSE
+		var/obj/item/stock_parts/cell = I
+		user.transferItemToLoc(I, loc)
+		cell.forceMove(src)
+		battery = cell
+		playsound(get_turf(src), 'sound/items/Deconstruct.ogg', 50, 1)
+		to_chat(user, "<span class='notice'>You slot \the [cell] inside \the [src]'s power supplier.</span>")
+		interact(user)
+		return TRUE
+	else
+		return ..()
+
+/obj/item/device/electronic_assembly/attack_self(mob/user)
+	if(!check_interactivity(user))
+		return
+	if(opened)
+		interact(user)
+
+	var/list/input_selection = list()
+	var/list/available_inputs = list()
+	for(var/obj/item/integrated_circuit/input/input in contents)
+		if(input.can_be_asked_input)
+			available_inputs.Add(input)
+			var/i = 0
+			for(var/obj/item/integrated_circuit/s in available_inputs)
+				if(s.name == input.name && s.displayed_name == input.displayed_name && s != input)
+					i++
+			var/disp_name= "[input.displayed_name] \[[input]\]"
+			if(i)
+				disp_name += " ([i+1])"
+			input_selection.Add(disp_name)
+
+	var/obj/item/integrated_circuit/input/choice
+	if(available_inputs)
+		var/selection = input(user, "What do you want to interact with?", "Interaction") as null|anything in input_selection
```

Comment:
No sanity checks after input.

Reply:
it's list input. What sanity check is needed there?

Reply:
Because the player may be 8 meters away, dead and burned to ash by the point the input window closes.

Reply:
added interactivity check after input.